aip/general/0236.md

```diff
@@ -0,0 +1,355 @@
+- If startPreview is called specifically on a resource which is already being previewed, then it **should** return an error message.
+- When startPreview is called, the system **should** set values for the log_name, and start_time in PreviewMetadata and set the state to ACTIVE.
+- If startPreview is called on an experiment with an empty set of rules, the deletion of that resource **must** be previewed.
+- startPreview is not idempotent and using it twice on the same resource **should** return an error.
+- If the API has an error, no fields in PreviewMetadata **should** change.
+- Etags **must** be used to guarantee resource freshness.
+
+#### stopPreview
+
+```proto
+rpc StopPreviewPolicyExperiment(StopPreviewPolicyExperimentRequest) returns (StopPreviewPolicyExperimentResponse) {
+  option (google.api.http) = {
+    post: "/v1/{name=policies/*/policyExperiments/*}:stopPreview"
+    body: "*"
+  }
+}
+
+message StopPreviewPolicyExperimentRequest {
+  string name = 1;
+  string etag = 2;
+}
+
+message StopPreviewPolicyExperimentResponse {
+  string name = 1;
+  string etag = 2;
+  string parent_etag = 3;
+  bool successful = 4;
+  string error_message = 5;
+}
+```
+- Admin flow API to stop previewing a resource **should** be implemented.
+  - Remains declarative friendly as only an output only field is modified.
+- If stopPreview is called specifically on a resource which is not actively being previewed, then it **should** return an error message.
+- When the API is called, PreviewMetadata.state **must** be set to INACTIVE and PreviewMetadata.stop_time **should** be set to the current time.
+- stopPreview is not idempotent, calling stop twice on the same resource **should** return an error.
+- If the API has an error, no fields in PreviewMetadata **should** change.
+- Etags **must** be used to guarantee resource freshness.
+
+#### commit
+
+A new API called commit **may** be implemented to promote the experimental configuration of a resource to the effective version and perform necessary cleanup. It will copy the fields-under-test from the experiment into the parent and then delete the experiment.
+
+This is a convenience API which is not declarative friendly. The alternative supported for declarative clients is to manually copy fields from an experiment onto the live policy then delete the experiment.
+
+```proto
+rpc CommitPolicyExperiment(CommitPolicyExperimentRequest) returns (CommitPolicyExperimentResponse) {
+  option (google.api.http) = {
+    post: "/v1/{name=policies/*/policyExperiments/*}:commit"
+    body: "*"
+  }
+}
+
+message CommitPolicyExperimentRequest {
+  string name = 1;
+  string etag = 2;
+  string parent_etag = 3; // optional.
+}
+
+message CommitPolicyExperimentResponse {
+  string parent_etag = 1;
+  bool successful = 2;
+  string error_message = 3;
+}
+```
+- Commit **must** atomically copy the fields-under-test from the experiment into the live policy then deletes the experiment.
```

Comment:
Deleting the resource as part of a successful commit call actually might be a little astonishing... Though I do understand the interest in having resources that can clean up after themselves when their job is finished...

Similar to above - would it be possible to consider two booleans that opt-in to the more automatic state changes the experiment is capable of?

`start_automatically: true` means to implicitly call `start` method after the experiment is create'ed or update'ed

`delete_automatically: true` means to implicitly call `delete` method after the experiment is cancel'ed or commit'ed

Reply:
I get the feeling that delete_automatically (which will be false by default) takes away the convenience of the method so I don't like the default behavior of delete_automatically being false.

(start_automatically has been addressed above)

Reply:
Fair point - it's just one of many ways that "hands-free" usage could be accomplished - and like you say those details can come later as long as the current design isn't painting into a corner in some way.